libs/Matchmaking/src/Peer/PeerDiscoveryAgent.cs

```diff
@@ -963,52 +974,68 @@ public IDiscoverySubscription Subscribe(string category)
             IReadOnlyDictionary<string, string> attributes = null,
             CancellationToken token = default)
         {
-            lock (this)
+            if (isDisposed_)
             {
-                if (server_ == null)
-                {
-                    server_ = new Server(transport_);
-                }
+                throw new ObjectDisposedException("PeerDiscoveryAgent");
+            }
+            if (server_ == null)
+            {
+                server_ = new Server(transport_);
             }
             AddRefToTransport();
             return server_.PublishAsync(category, connection, options_.ResourceExpirySec, attributes, token);
         }
 
         private void AddRefToTransport()
         {
-            int newRefCount = Interlocked.Increment(ref transportRefCount_);
-            if (newRefCount == 1)
+            lock (transportStartStopLock_)
             {
-                transport_.Start();
+                if (transportRefCount_ == 0)
+                {
+                    transport_.Start();
```

Comment:
Beyond scope of this review, but I'm finding myself questioning why the service is responsible for starting/stopping the transport.

Reply:
We want to react to when the agent starts/stops listening, so maybe it's fine to move the Started/Stopped event to the agent - then we won't need to start/stop the transport. I don't think keeping the socket/threads active will make a lot of difference.

IIRC the events were put in the transport because they were intuitively more related to the transport than to the agent, but as things are now a transport is exclusively used by an agent, so it doesn't make much difference.